### ISSUE TITLE
Avoid executing queries which are known to return no results

### ISSUE THREAD

**Issue by Varorbc** [aspnet/EntityFrameworkCore#31405]
The query in the example is actually meaningless, it should directly return an empty list instead of querying the database

```
using System.ComponentModel.DataAnnotations;

var services = new ServiceCollection();
services.AddDbContext<TestContext>();
var serviceProvider = services.BuildServiceProvider();

var testContext = serviceProvider.GetRequiredService<TestContext>();
var ids = new List<Guid>();
var test = testContext.Test1.Where(a => ids.Contains(a.Id)).ToList();

public partial class Test1
{
    [Key]
    public Guid Id { get; set; }

    [StringLength(50)]
    public string Name { get; set; } = null!;
}

public class TestContext : DbContext
{
    public TestContext()
    {
    }

    public TestContext(DbContextOptions<TestContext> options)
        : base(options)
    {
    }

    public virtual DbSet<Test1> Test1 { get; set; }


    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        _ = optionsBuilder.UseSqlServer("");
    }
}
```

**Comment by roji** [comment]
@Varorbc this has been discussed before (I can't find the issue at the moment).

Although it's true we could do this, it isn't trivial, and it's very easy for users to simply perform the check themselves and not do the query. In other words, it's an optimization that can happen outside of EF just as well as inside - so we'd give this quite a low priority.

**Comment by stevendarby** [comment]
Consider something like an OData filter string which is supplied externally by a user. We leave it to OData to translate the string into a filter expression and we leave it to EF to parse the expression into a SQL query. It's not trivial to inspect the filter's validity at any stage without doing the parsing & translating that we'd otherwise leave to those libraries. So if EF optimised for us here, that'd be great!

**Comment by roji** [comment]
@stevendarby sure, I do understand how that can be useful. I'd argue that asking the OData layer to do this is just as reasonable as asking EF to do it (though that's not a reason for us not to do it).

But stepping, our overall experience suggests that optimizations on the LINQ query itself are better done upstream, at the source actually producing that LINQ query, rather than trying to compensate for them downstream in EF. For example, we occasionally get issues where some LINQ query comes in with duplicated clauses (because an automated layer such as OData produces them), and the user asks us to pattern-match, identify those duplications and optimize them away. That can become unreasonably complicated and unfeasible really quickly.

But I do agree it could be a good idea to have at least some infrastructure for not sending a query to the database when we know there will be no results.

**Comment by stevendarby** [comment]
> @stevendarby sure, I do understand how that can be useful. I'd argue that asking the OData layer to do this is just as reasonable as asking EF to do it (though that's not a reason for us not to do it).

Hmm, but doesn’t OData just modify the IQueryable whereas EF is the query provider whose job is to materialise that queryable into results- and it could decide to do that by running a 1=0 query or just returning an empty list. I can’t think what OData could do differently in terms of modifying the queryable other than maybe throw which isn’t great.

**Comment by roji** [comment]
@stevendarby I don't follow... OData is the one building a LINQ expression tree (from user URL input) and passing it to EF; and it's building an expression tree that can mathematically have no results. Wouldn't it make *more* sense for OData to not invoke EF at all in this case?

**Comment by stevendarby** [comment]
@roji AFAIK OData _doesn’t_ invoke EF / materialise results itself. It merely modifies the IQueryable by adding Where filters etc.

What happens after that may depend on exactly what bits of the library you’re using: it could be this IQueryable is passed back to the developer to invoke eg ToList themselves, or it’s passed on to the ASP.NET framework to enumerate over the IQueryable when writing the response, etc. at this point, if the IQueryable was created with EF then EF is still the query provider even after passing through OData and so still has the job of materialising the results. 

In any case, probably worth me stepping back to say this request is just a nice to have in my opinion :D Just wanted to say it’s not always simple for the developer to do this check themselves. I take the general point that optimisations are often better done upstream though.

**Comment by roji** [comment]
I'm definitely no OData expert, but in my mental model OData is the one *creating* the IQueryable rather than modifying it; otherwise which component is the one actually creating it? Whoever is creating it - or even modifying it - could recognize at the point where the "impossible" filter is integrated, and simply pass back an empty IQueryable or similar, I think. That would even have the advantage of optimizing the query away for other, non-EF queryable providers.

But I think we're overall in agreement - this would indeed be a nice-to-have optimization in EF - but unfortunately this isn't trivial implementation-wise. It's specifically a bit tricky that the "nonsense" part comes from a parameter value (the list of IDs in this case); EF generally does most of its query compilation while abstracting away parameters (so that the query can be reused regardless of actual values). So contrary to cases where we can know there won't be any results in compilation itself (e.g. the user writing `.Where(false)`), in this case we'd need to insert some sort of parameter check into the generated query executor, to happen every time the query is invoked. We don't have anything like that at the moment.

**Comment by stevendarby** [comment]
> I'm definitely no OData expert, but in my mental model OData is the one creating the IQueryable rather than modifying it. 

Ah no, OData merely acts on an existing data source which it’s agnostic about. So if you’re using EF, you’ll usually start with something like `context.Set<Blog>()` and hand that over to OData to apply filters and other operations.

> Whoever is creating it - or even modifying it - could recognize at the point where the "impossible" filter is integrated, and simply pass back an empty IQueryable or similar

I think it would be bad form for it return an entirely new IQueryable with a different query provider - e.g. `new List<Blog>.AsQueryable()`. It would mean that the returned IQueryable’s provider is sometimes still EF (when it’s a “possible” filter that needs to go to the DB) and sometimes not (when it’s an “impossible” filter). The developer might use EF’s ToListAsync() extension on it and find it throws in the latter case - but the developer can’t easily judge when this is going to be the case.

Therefore I think it’s correct that OData can only append a filter to the existing queryable- the best it can do is detect “impossible” filters and optimise them down to something like a `Where(false)`, but that stills end up with EF deciding what to do with it. 

> But I think we're overall in agreement - this would indeed be a nice-to-have optimization in EF - but unfortunately this isn't trivial implementation-wise. It's specifically a bit tricky that the "nonsense" part comes from a parameter value (the list of IDs in this case); EF generally does most of its query compilation while abstracting away parameters (so that the query can be reused regardless of actual values). So contrary to cases where we can know there won't be any results in compilation itself (e.g. the user writing .Where(false)), in this case we'd need to insert some sort of parameter check into the generated query executor, to happen every time the query is invoked. We don't have anything like that at the moment.

Ah this where I start to think we may be talking about different things. EF already quite often detects “impossible” filters and executes a SQL query containing `WHERE 1 = 0`. I assumed OP’s issue was one such case. I was thinking this is only about handling those cases differently - rather than run with `WHERE 1 = 0`, don’t run it at all. That doesn’t seem too complex in my mind. But maybe this is about something different?

**Comment by stevendarby** [comment]
Ah, @roji, part of my confusion comes from the change in behaviour with EF 8...

EF 7:
```SQL
SELECT [t].[Id], [t].[Name]
FROM [Test1] AS [t]
WHERE 0 = 1
```
EF 8:
```SQL
SELECT [t].[Id], [t].[Name]
FROM [Test1] AS [t]
WHERE [t].[Id] IN (
    SELECT [i].[value]
    FROM OPENJSON(@__ids_0) WITH ([value] uniqueidentifier '$') AS [i]
)
```

Nonetheless there may still be some non `Contains` type query filters that EF still translates to `WHERE 1 = 0` - so maybe these could still be the target of an optimisation?

**Comment by roji** [comment]
> Ah no, OData merely acts on an existing data source which it’s agnostic about. So if you’re using EF, you’ll usually start with something like context.Set<Blog>() and hand that over to OData to apply filters and other operations.

Of course, but the point is that apart from the query root (the DbSet), everything else is generated by EF - notably the filter which makes the query impossible.

> I think it would be bad form for it return an entirely new IQueryable with a different query provider - e.g. new List<Blog>.AsQueryable(). It would mean that the returned IQueryable’s provider is sometimes still EF (when it’s a “possible” filter that needs to go to the DB) and sometimes not (when it’s an “impossible” filter). The developer might use EF’s ToListAsync() extension on it and find it throws in the latter case - but the developer can’t easily judge when this is going to be the case.

That's an interesting point... I'd have to check, but IIRC ToListAsync only requires that the IQueryable also implement IAsyncEnumerable (in addition to implementing IEnumerable, from which IQueryable derives). So I don't **think** there's any problem with returning a different IQueryable implementation as long as it also implement IAsyncEnumerable.

In that sense IQueryable should largely maintain the intended abstraction, i.e. its consumer isn't supposed to care about whether an IQueryable is backed by EF or any other LINQ provider. The sync/async detail does leak out a bit (and the story for queryable and async is generally not great), but that's all a result of async having been introduced later, after IQueryable was already there.

> Ah this where I start to think we may be talking about different things. EF already quite often detects “impossible” filters and executes a SQL query containing WHERE 1 = 0.

Yes, but there's a big distinction between cases where the impossibility is part of the query itself, and where it's a result of a specific parameter value (as is the case here). The former is much easier: we know during translation of the query tree - regardless of what any parameter values are - that the query won't return anything. Even there, returning an empty IEnumerable (rather than actually executing with `WHERE 1 = 0`) would require some work; and those cases are particularly senseless to optimize, since it really is a silly query by construction.

> Ah, @roji, part of my confusion comes from the change in behaviour with EF 8...

Yes, that's indeed very relevant - hopefully it's clearer why detecting and optimizing this away isn't trivial. The point with the EF 8.0 optimization of contains was precisely to **stop** varying the SQL based on parameter values (since that is problematic for plan caching and other things). Contains now behaves just like any other operator, with the list getting properly parameterized (as a JSON array) rather than getting integrated into the SQL itself.

**Comment by stevendarby** [comment]
@roji, good point about IQueryable’s intended abstraction. That and the issue with optimising in EF are much clearer in my mind now. Sorry to use up so much of your time but I really valued the discussion :)

**Comment by roji** [comment]
Oh no worries at all, these discussions are very useful for me too - always happy to have them!

**Comment by roji** [comment]
Putting this in the backlog as a potential optimization we could do, but which seems rather low-value compared to the work that's likely required.

Note that implementation-wise, there are two variants of queries which are known to return nothing:

* The query always returns nothing, because of its structure (e.g. `Where(_ => false)`). This can be detected during (1st-phase compilation), and  could be implemented by having a QueryCompilationContext facility for saying that the query returns nothing, and that would trigger returning a no-op query executor.
* The query only returns nothing for specific parameter values. This is the case with @Varorbc's original scenario:

```c#
var ids = new List<Guid>();
var test = testContext.Test1.Where(a => ids.Contains(a.Id)).ToList();
```

* In this case, since the `ids` list is fully parameterized (starting with EF 8), we can only know that query database execution is unnecessary in the 2nd part of the query pipeline. That likely requires a separate mechanism for avoiding execution etc.
